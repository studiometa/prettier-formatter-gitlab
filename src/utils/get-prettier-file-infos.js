import { readFileSync } from 'node:fs';
import { relative, dirname } from 'node:path';
import {
  resolveConfig,
  resolveConfigFile,
  getFileInfo as prettierGetFileInfo,
  check,
  format,
} from 'prettier';

/**
 * @typedef {import('../types.d.ts').FileInfo} FileInfo
 */

/**
 * Get Prettier file informations.
 * @param {string} filePath The absolute path to the file.
 * @returns {Promise<FileInfo>} An object with Prettier informations.
 */
async function getFileInfos(filePath) {
  const input = readFileSync(filePath, 'utf8').toString();
  const [options, configPath] = await Promise.all([
    resolveConfig(filePath),
    resolveConfigFile(filePath),
  ]);
  const filename = relative(dirname(configPath), filePath);
  const infos = await prettierGetFileInfo(filePath, options);

  if (!options.parser) {
    options.parser = infos.inferredParser;
  }

  try {
    const [isFormatted, output] = await Promise.all([
      check(input, options),
      format(input, options),
    ]);

    return {
      filename,
      input,
      output,
      isFormatted,
    };
  } catch (error) {
    return {
      filename,
      error,
    };
  }
}

const cache = new Map();

/**
 * Get Prettier information for a given file.
 * @param   {string} filePath The absolute path to the file.
 * @returns {Promise<FileInfo>}
 */
export async function getPrettierFileInfos(filePath) {
  if (!cache.has(filePath)) {
    const fileInfo = await getFileInfos(filePath);
    cache.set(filePath, fileInfo);
  }

  return cache.get(filePath);
}
