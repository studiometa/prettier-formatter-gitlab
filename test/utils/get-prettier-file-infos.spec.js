import { describe, it, expect } from 'bun:test';
import { resolve } from 'node:path';
import { getPrettierFileInfos } from '../../src/utils/get-prettier-file-infos.js';

describe('The getPrettierFileInfos function', () => {
  it('should return information about a given file', async () => {
    const filepath = resolve(import.meta.dirname, '../__stubs__/dirty.js');
    const infos = await getPrettierFileInfos(filepath);
    expect(infos).toMatchSnapshot();
  });

  it('should handle files with syntax errors', async () => {
    const filepath = resolve(import.meta.dirname, '../__stubs__/syntax-error.js');
    const infos = await getPrettierFileInfos(filepath);
    expect(infos).toMatchSnapshot();
  });
});
