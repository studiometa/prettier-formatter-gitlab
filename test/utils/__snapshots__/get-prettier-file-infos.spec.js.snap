// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`The getPrettierFileInfos function should return information about a given file 1`] = `
{
  "filename": "test/__stubs__/dirty.js",
  "input": 
"/* eslint-disable */







function foo( arg  ) {


  const data = arg.split(',');




  return data;
}
"
,
  "isFormatted": false,
  "output": 
"/* eslint-disable */

function foo(arg) {
  const data = arg.split(',');

  return data;
}
"
,
}
`;

exports[`The getPrettierFileInfos function should handle files with syntax errors 1`] = `
{
  "error": [SyntaxError: Unexpected token, expected "(" (3:13)
\x1B[0m \x1B[90m 1 |\x1B[39m \x1B[90m/* eslint-disable */\x1B[39m
 \x1B[90m 2 |\x1B[39m
\x1B[31m\x1B[1m>\x1B[22m\x1B[39m\x1B[90m 3 |\x1B[39m \x1B[36mfunction\x1B[39m foo\x1B[33m#\x1B[39m\x1B[33m#\x1B[39m\x1B[33m#\x1B[39m\x1B[33m#\x1B[39m\x1B[33m#\x1B[39m( arg  ) {
 \x1B[90m   |\x1B[39m             \x1B[31m\x1B[1m^\x1B[22m\x1B[39m
 \x1B[90m 4 |\x1B[39m   \x1B[36mreturn\x1B[39m arg\x1B[33m;\x1B[39m
 \x1B[90m 5 |\x1B[39m }
 \x1B[90m 6 |\x1B[39m\x1B[0m],
  "filename": "test/__stubs__/syntax-error.js",
}
`;
